jsdom-based-scraping copy/index.md

```diff
@@ -0,0 +1,207 @@
+
+
+We will start this approach by creating a function that will create the URL for the API call for us and make the call and get the data.
+
+```js
+export const createStartUrls = (input) => {
+    const {
+        days = '7',
+        country = '',
+        resultsLimit = 100,
+        industry = '',
+        isNewToTop100,
+    } = input;
+
+    const filterBy = isNewToTop100 ? 'new_on_board' : '';
+    return [{
+        url: `https://ads.tiktok.com/creative_radar_api/v1/popular_trend/hashtag/list?page=1&limit=50&period=${days}&country_code=${country}&filter_by=${filterBy}&sort_by=popular&industry_id=${industry}`,
+        headers: {
+           // required headers
+        },
+        userData: { resultsLimit },
+    }];
+};
+```
+
+In the above function, we create the start url for the API call that include various parameters as we talked about earlier. After creating the URL according to the parameters it will call the `creative_radar_api` and fetch all the results.
+
+But it won’t work until we get the headers. So let’s create a function that will first create a session using sessionPool and proxyConfiguration.
+
+```js
+export const createSessionFunction = async (sessionPool, proxyConfiguration) => {
+    const proxyUrl = await proxyConfiguration.newUrl(Math.random().toString());
+    const url = 'https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pad/en';
+    // need url with data to generate token
+    const response = await gotScraping({ url, proxyUrl });
+    const headers = await getApiUrlWithVerificationToken(response.body.toString(), url);
+    if (!headers) {
+        throw new Error(`Token generation blocked`);
+    }
+    log.info(`Generated API verification headers`, Object.values(headers));
+    return new Session({
+        userData: {
+            headers,
+        },
+        sessionPool,
+    });
+};
+```
+In this function, the main goal is to call `https://ads.tiktok.com/business/creativecenter/inspiration/popular/hashtag/pad/en
+` and get headers in return. To get the headers we are using `getApiUrlWithVerificationToken` function.
+
+Let’s see what’s there in this function:
+
+
+```js
+const getApiUrlWithVerificationToken = async (body, url) => {
+    log.info(`Getting API session`);
+    const virtualConsole = new VirtualConsole();
+    const { window } = new JSDOM(body, {
+        url,
+        contentType: 'text/html',
+        runScripts: 'dangerously',
+        resources: 'usable' || new CustomResourceLoader(),
+        // ^ 'usable' faster than custom and works wo canvas
+        pretendToBeVisual: false,
+        virtualConsole,
+    });
+    virtualConsole.on('error', () => {
+        // ignore errors cause by fake XMLHttpRequest
+    });
+    // might need stubs for custom resources https://github.com/apify/crawlee/blob/f35b80501d1f926935fb75a875009a2cfabd31d5/packages/jsdom-crawler/src/internals/jsdom-crawler.ts#L232
```

Comment:
remove "might need stubs for custom resources" - not needed so far, therefore just confusing